source/material_model/rheology/diffusion_creep.cc

```diff
@@ -39,52 +39,91 @@ namespace aspect
+
         // Power law creep equation
         //    viscosity = 0.5 * A^(-1/n) * d^(m/n) * exp((E + P*V)/(nRT))
         // A: prefactor,
         // d: grain size, m: grain size exponent, E: activation energy, P: pressure,
         // V; activation volume, R: gas constant, T: temperature.
-        const double viscosity_diffusion = 0.5 / prefactors_diffusion_composition *
-                                           std::exp((activation_energies_diffusion_composition +
-                                                     pressure*activation_volumes_diffusion_composition)/
+        const double viscosity_diffusion = 0.5 / p.prefactor *
+                                           std::exp((p.activation_energy +
+                                                     pressure*p.activation_volume)/
                                                     (constants::gas_constant*temperature)) *
-                                           std::pow(grain_size, grain_size_exponents_diffusion_composition);
+                                           std::pow(grain_size, p.grain_size_exponent);
 
         return viscosity_diffusion;
       }
 
 
+      template <int dim>
+      std::pair<double, double>
+      DiffusionCreep<dim>::compute_strain_rate_and_derivative (const double stress,
+                                                               const double pressure,
+                                                               const double temperature,
+                                                               const DiffusionCreepParameters creep_parameters) const
+      {
+        // Power law creep equation
+        //   edot_ii_partial = A * stress^n * d^-m * exp(-(E + P*V)/(RT))
+        //   d(edot_ii_partial)/d(stress) = A * n * stress^(n-1) * d^-m * exp(-(E + P*V)/(RT))
+        // A: prefactor, edot_ii_partial: square root of second invariant of deviatoric strain rate tensor attributable to the creep mechanism,
+        // d: grain size, m: grain size exponent, E: activation energy, P: pressure,
+        // V; activation volume, R: gas constant, T: temperature.
+        // For diffusion creep, n = 1 (strain rate is linearly dependent on stress).
+        const double dstrain_rate_dstress_diffusion = creep_parameters.prefactor *
+                                                      std::pow(grain_size, -creep_parameters.grain_size_exponent) *
+                                                      std::exp(-(creep_parameters.activation_energy + pressure*creep_parameters.activation_volume)/
+                                                               (constants::gas_constant*temperature));
+
+        const double strain_rate_diffusion = stress * dstrain_rate_dstress_diffusion;
+
+        const std::pair<double, double> strain_rate_and_derivative (strain_rate_diffusion, dstrain_rate_dstress_diffusion);
+        return strain_rate_and_derivative;
```

Comment:
this can be somewhat simplified in a rare case of shorter *and* simpler to understand :-)

```suggestion
        return std::make_pair(strain_rate_diffusion, dstrain_rate_dstress_diffusion);
```

Reply:
aha!